engine/src/main/java/org/terasology/world/chunks/pipeline/ChunkProcessingPipeline.java

```diff
@@ -0,0 +1,206 @@
+// Copyright 2020 The Terasology Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+package org.terasology.world.chunks.pipeline;
+
+import com.google.common.collect.Sets;
+import org.joml.Vector3i;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.utilities.concurrency.TaskMaster;
+import org.terasology.world.chunks.Chunk;
+import org.terasology.world.chunks.pipeline.tasks.ChunkTaskListenerWrapper;
+
+import java.util.Comparator;
+import java.util.Deque;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Manages execution of chunk tasks on a queue.
+ * <p>
+ * {@link ChunkTask}s are executing in background threads.
+ * <p>
+ * {@link ChunkTask}s are executing by priority via {@link Comparable}.
+ * <p>
+ * {@link Chunk}s will processing on stages {@link ChunkProcessingPipeline#addStage}
+ */
+public class ChunkProcessingPipeline implements ChunkTaskListener {
+    private static final int NUM_TASK_THREADS = 8;
```

Comment:
This reminds me, I want this thread pool size to be configurable!

(That can be out-of-scope for this PR, this is just a note marking the place to do it.)